File: sdk/ruby/lib/velocity_sdk/activity_options.rb
# frozen_string_literal: true

module VelocitySdk
  # Configuration options for activity execution.
  #
  # Controls timeouts, retry behaviour, and task queue routing for
  # individual activity invocations within a workflow.
  #
  # @example
  #   opts = VelocitySdk::ActivityOptions.new
  #     .start_to_close_timeout_ms(5_000)
  #     .retry_max_attempts(3)
  #     .task_queue('activity-workers')
  class ActivityOptions
    # @return [Integer] Start-to-close timeout in milliseconds.
    attr_reader :start_to_close_timeout_ms

    # @return [Integer] Schedule-to-close timeout in milliseconds.
    attr_reader :schedule_to_close_timeout_ms

    # @return [Integer] Heartbeat timeout in milliseconds (0 = disabled).
    attr_reader :heartbeat_timeout_ms

    # @return [String] Task queue for activity dispatch.
    attr_reader :task_queue

    # @return [Integer] Maximum number of retry attempts.
    attr_reader :retry_max_attempts

    # @return [Float] Retry backoff coefficient.
    attr_reader :retry_backoff_coefficient

    # @return [Integer] Initial retry interval in milliseconds.
    attr_reader :retry_initial_interval_ms

    # Create a new options builder with defaults.
    def initialize
      @start_to_close_timeout_ms = 10_000
      @schedule_to_close_timeout_ms = 60_000
      @heartbeat_timeout_ms = 0
      @task_queue = 'default'
      @retry_max_attempts = 1
      @retry_backoff_coefficient = 2.0
      @retry_initial_interval_ms = 100
    end

    # Create options with all defaults.
    # @return [ActivityOptions]
    def self.defaults
      new
    end

    # Set the start-to-close timeout in milliseconds.
    # @param value [Integer]
    # @return [self]
    def start_to_close_timeout_ms(value)
      @start_to_close_timeout_ms = [0, value].max
      self
    end

    # Set the schedule-to-close timeout in milliseconds.
    # @param value [Integer]
    # @return [self]
    def schedule_to_close_timeout_ms(value)
      @schedule_to_close_timeout_ms = [0, value].max
      self
    end

    # Set the heartbeat timeout in milliseconds.
    # @param value [Integer]
    # @return [self]
    def heartbeat_timeout_ms(value)
      @heartbeat_timeout_ms = [0, value].max
      self
    end

    # Set the task queue for activity dispatch.
    # @param value [String]
    # @return [self]
    def task_queue(value)
      @task_queue = value
      self
    end

    # Set the maximum number of retry attempts.
    # @param value [Integer]
    # @return [self]
    def retry_max_attempts(value)
      @retry_max_attempts = [1, value].max
      self
    end

    # Set the retry backoff coefficient.
    # @param value [Float]
    # @return [self]
    def retry_backoff_coefficient(value)
      @retry_backoff_coefficient = [1.0, value].max
      self
    end

    # Set the initial retry interval in milliseconds.
    # @param value [Integer]
    # @return [self]
    def retry_initial_interval_ms(value)
      @retry_initial_interval_ms = [0, value].max
      self
    end

    # Convert to a Hash for serialisation.
    # @return [Hash]
    def to_h
      {
        start_to_close_timeout_ms: @start_to_close_timeout_ms,
        schedule_to_close_timeout_ms: @schedule_to_close_timeout_ms,
        heartbeat_timeout_ms: @heartbeat_timeout_ms,
        task_queue: @task_queue,
        retry_max_attempts: @retry_max_attempts,
        retry_backoff_coefficient: @retry_backoff_coefficient,
        retry_initial_interval_ms: @retry_initial_interval_ms,
      }
    end

    # @return [String]
    def inspect
      "#<VelocitySdk::ActivityOptions tq=#{@task_queue} timeout=#{@start_to_close_timeout_ms}ms retries=#{@retry_max_attempts}>"
    end
  end
end

File: sdk/php/src/Activity/ActivityOptions.php
<?php

declare(strict_types=1);

namespace Velocity\SDK\Activity;

/**
 * Configuration options for activity execution.
 *
 * Controls timeouts, retry behaviour, and task queue routing for
 * individual activity invocations within a workflow.
 *
 * Usage:
 *     $opts = ActivityOptions::new()
 *         ->withStartToCloseTimeoutMs(5_000)
 *         ->withRetryMaxAttempts(3)
 *         ->withTaskQueue('activity-workers');
 */
class ActivityOptions
{
    private int $startToCloseTimeoutMs = 10_000;
    private int $scheduleToCloseTimeoutMs = 60_000;
    private int $heartbeatTimeoutMs = 0;
    private string $taskQueue = 'default';
    private int $retryMaxAttempts = 1;
    private float $retryBackoffCoefficient = 2.0;
    private int $retryInitialIntervalMs = 100;

    private function __construct()
    {
    }

    /** Create a new options builder. */
    public static function new(): self
    {
        return new self();
    }

    /** Create options with all defaults. */
    public static function defaults(): self
    {
        return new self();
    }

    /** Set the start-to-close timeout in milliseconds. */
    public function withStartToCloseTimeoutMs(int $ms): self
    {
        $this->startToCloseTimeoutMs = max(0, $ms);
        return $this;
    }

    /** Set the schedule-to-close timeout in milliseconds. */
    public function withScheduleToCloseTimeoutMs(int $ms): self
    {
        $this->scheduleToCloseTimeoutMs = max(0, $ms);
        return $this;
    }

    /** Set the heartbeat timeout in milliseconds (0 = disabled). */
    public function withHeartbeatTimeoutMs(int $ms): self
    {
        $this->heartbeatTimeoutMs = max(0, $ms);
        return $this;
    }

    /** Set the task queue for activity dispatch. */
    public function withTaskQueue(string $taskQueue): self
    {
        $this->taskQueue = $taskQueue;
        return $this;
    }

    /** Set the maximum number of retry attempts. */
    public function withRetryMaxAttempts(int $attempts): self
    {
        $this->retryMaxAttempts = max(1, $attempts);
        return $this;
    }

    /** Set the retry backoff coefficient. */
    public function withRetryBackoffCoefficient(float $coefficient): self
    {
        $this->retryBackoffCoefficient = max(1.0, $coefficient);
        return $this;
    }

    /** Set the initial retry interval in milliseconds. */
    public function withRetryInitialIntervalMs(int $ms): self
    {
        $this->retryInitialIntervalMs = max(0, $ms);
        return $this;
    }

    // ─── Accessors ────────────────────────────────────────────────────────

    public function getStartToCloseTimeoutMs(): int { return $this->startToCloseTimeoutMs; }
    public function getScheduleToCloseTimeoutMs(): int { return $this->scheduleToCloseTimeoutMs; }
    public function getHeartbeatTimeoutMs(): int { return $this->heartbeatTimeoutMs; }
    public function getTaskQueue(): string { return $this->taskQueue; }
    public function getRetryMaxAttempts(): int { return $this->retryMaxAttempts; }
    public function getRetryBackoffCoefficient(): float { return $this->retryBackoffCoefficient; }
    public function getRetryInitialIntervalMs(): int { return $this->retryInitialIntervalMs; }

    /**
     * Convert to an associative array for serialisation.
     *
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        return [
            'start_to_close_timeout_ms' => $this->startToCloseTimeoutMs,
            'schedule_to_close_timeout_ms' => $this->scheduleToCloseTimeoutMs,
            'heartbeat_timeout_ms' => $this->heartbeatTimeoutMs,
            'task_queue' => $this->taskQueue,
            'retry_max_attempts' => $this->retryMaxAttempts,
            'retry_backoff_coefficient' => $this->retryBackoffCoefficient,
            'retry_initial_interval_ms' => $this->retryInitialIntervalMs,
        ];
    }
}
